### PR TITLE
MGMT-2187: Optimize image upload by using UploadPartCopy

### PR DESCRIPTION
**Pull request by avishayt**

### CODE REVIEW

**Comment by avishayt** [comment]
Successfully ran assisted-test-infra and manually tested onprem (podman).

File: assisted-iso-create/main.go [hunk 1]
```diff
@@ -1,74 +1,22 @@
...
-func uploadCreatedISOToS3(s3Client *s3wrapper.S3Client, isoFile, isoObjectName string, log *logrus.Logger) error {
-	ctx := context.Background()
-	err := s3Client.UploadFile(ctx, isoFile, isoObjectName)
-	if err != nil {
-		log.Errorf("Failed to upload file %s as object %s", isoFile, isoObjectName)
-		return err
-	}
-	_, err = s3Client.UpdateObjectTimestamp(ctx, isoObjectName)
-	return err
+	WorkDir     string `envconfig:"WORK_DIR"`
```
Comment (filanov) [inline assisted-iso-create/main.go:12]: I don't think it's used.

File: assisted-iso-create/main.go [hunk 2]
```diff
@@ -79,31 +27,21 @@ func main() {
...
-			log.Fatal("failed to create S3 client, ", err)
-		}
-
-		err = uploadCreatedISOToS3(s3Client, createImageFile, Options.ImageName, log)
-		if err != nil {
-			log.Fatal(err.Error())
-		}
-	} else {
-		log.Println("Using local storage for image")
+	_, err = s3Client.UpdateObjectTimestamp(ctx, Options.ImageName)
```
Comment (filanov) [inline assisted-iso-create/main.go:41]: When uploading a base image should we overwrite the existing one or should we create a new one?

File: cmd/main.go [hunk 3]
```diff
@@ -287,9 +287,10 @@ func main() {
 	if Options.DeployTarget == deploymet_type_k8s {
 		go func() {
 			defer apiEnabler.Enable()
-			//Run first ISO dummy for image pull, this is done so that the image will be pulled and the api will take less time.
-			// blocking function that can take a long time.
-			bminventory.GenerateDummyISOImage(log, generator, eventsHandler)
+			// Upload the live image which will serve as a basis for user-generated images.
+			if err = generator.UploadBaseISO(); err != nil {
```
Comment (filanov) [inline cmd/main.go:291]: On each update we will have 3 replicas that will upload the iso to the same place. maybe we should consider leader election? or check if we already have a correct version in s3? 
Maybe it's not a real issue for now and can be resolved in another PR.

File: pkg/job/local_job.go [hunk 4]
```diff
@@ -12,12 +12,12 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/openshift/assisted-service/internal/common"
-	"github.com/openshift/assisted-service/internal/events"
 	"github.com/openshift/assisted-service/pkg/generator"
 	logutil "github.com/openshift/assisted-service/pkg/log"
 	"github.com/sirupsen/logrus"
 )
 
+//go:generate mockgen -source=local_job.go -package=job -destination=mock_local_job.go
```
Comment (filanov) [inline pkg/job/local_job.go:20]: i think that 
```
//go:generate mockgen -package=job -destination=mock_local_job.go . LocalJob
``` 
will work as well

File: pkg/s3wrapper/client.go [hunk 5]
```diff
@@ -23,12 +26,15 @@ import (
...
 	"github.com/aws/aws-sdk-go/service/s3/s3iface"
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/cavaliercoder/go-cpio"
 	"github.com/go-openapi/swag"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
 const awsEndpointSuffix = ".amazonaws.com"
+const baseObjectName = "livecd.iso"
```
Comment (filanov) [inline pkg/s3wrapper/client.go:36]: I think that i already saw this defined somewhere? probably should be in one place?

File: pkg/s3wrapper/client.go [hunk 6]
```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
...
+
+	res := bytes.Compare(headerString[0:8], []byte("coreiso+"))
+	if res != 0 {
+		return 0, 0, errors.New("Could not find magic string in object header")
+	}
+
+	offset := int64(binary.LittleEndian.Uint64(headerString[8:16]))
+	length := int64(binary.LittleEndian.Uint64(headerString[16:24]))
+
+	// For now we assume that the embedded area is less than 5MB, which is the minimum S3 part size
```
Comment (YuviGold) [inline pkg/s3wrapper/client.go:280]: Did you meant maximum?

File: pkg/s3wrapper/client.go [hunk 7]
```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
...
+	if length > int64(fiveMB) {
+		return 0, 0, errors.New("ISO embedded area larger than what is currently supported")
```
Comment (YuviGold) [inline pkg/s3wrapper/client.go:282]: `is larger` / `is greater`

File: pkg/s3wrapper/client.go [hunk 8]
```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
...
+	}
+	return offset, length, nil
+}
+
+func (c *S3Client) uploadPartCopy(log logrus.FieldLogger, partNum *int64, uploadID *string, sourceObjectKey string, destObjectKey string,
```
Comment (YuviGold) [inline pkg/s3wrapper/client.go:287]: Shouldn't be a pointer. You print and use the address of the pointer in this function.

File: pkg/s3wrapper/client.go [hunk 9]
```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
...
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to start upload for %s", objectName)
+		log.Error(err)
+		return err
+	}
+	uploadID := multiOut.UploadId
+	var completedParts []*s3.CompletedPart
+
+	// First part: copy the first part of the live ISO, until the embedded area
+	completedPart, err := c.uploadPartCopy(log, aws.Int64(1), uploadID, baseObjectName, objectName, 0, areaOffsetBytes-1)
```
Comment (YuviGold) [inline pkg/s3wrapper/client.go:201]: The function gets `*int64` as `partNumber`.
Idk what is `aws.Int64` type but I think the usage should be either `aws.Int64` in both places or `int64` accordingly (better IMO).

File: pkg/s3wrapper/client.go [hunk 10]
```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
...
+		return nil, err
+	}
+	if err := gzipWriter.Close(); err != nil {
+		err = errors.Wrapf(err, "Failed to gzip ignition config")
+		log.Error(err)
+		return nil, err
+	}
+
+	if int64(len(compressedBuffer.Bytes())) > areaLengthBytes {
+		err := errors.New(fmt.Sprintf("Ignition is too long to be embedded (%d > %d)", len(archiveBuffer.Bytes()), areaLengthBytes))
```
Comment (YuviGold) [inline pkg/s3wrapper/client.go:340]: Replace `archiveBuffer` with `compressedBuffer`
```golang
		err := errors.New(fmt.Sprintf("Ignition is too long to be embedded (%d > %d)", len(compressedBuffer.Bytes()), areaLengthBytes))

```

File: pkg/s3wrapper/client.go [hunk 11]
```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
...
+	completedPartCopy, err := c.client.UploadPart(&s3.UploadPartInput{
+		Bucket:        aws.String(c.cfg.S3Bucket),
+		Key:           aws.String(objectName),
+		PartNumber:    partNum,
+		UploadId:      uploadID,
+		Body:          bytes.NewReader(origContents),
+		ContentLength: aws.Int64(contentLength),
+	})
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to upload ignition for file %s", objectName)
```
Comment (YuviGold) [inline pkg/s3wrapper/client.go:357]: Maybe we should provide more metadata? `uploadID`, `S3Bucket`, `contentLength`..

File: pkg/s3wrapper/client.go (hunk 9 above)
Comment (filanov) [reply pkg/s3wrapper/client.go:201]: `aws.Int64` is the same as `swag.Int64` i think it's fine to use aws infra for aws functions.

**Comment by avishayt** [comment]
`Ran 69 of 69 Specs in 621.840 seconds
SUCCESS! -- 69 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestSubsystem (621.85s)
PASS
`

Test infra passed as well

**Review by filanov — APPROVED** [review]
